test/extended/util/disruption/disruption.go

```diff
@@ -197,71 +201,79 @@ func (cma *chaosMonkeyAdapter) Test(sem *chaosmonkey.Semaphore) {
-					Message: test.Failure,
-					Value:   test.Failure,
-				})
+				tc.FailureOutput = &junitapi.FailureOutput{Message: test.Failure}
 			}
-			ts.TestCases = append(ts.TestCases, testCase)
+			ts.TestCases = append(ts.TestCases, tc)
 			continue
 		}
 
-		filePath := filepath.Join(framework.TestContext.ReportDir, fmt.Sprintf("%s_%s_%s.json", summary.SummaryKind(), filesystemSafeName(tc.Name), now.Format(time.RFC3339)))
+		// TODO: this is writing out Flake_[testname]_[timestamp].json files with content that is just: {"type":"Flake"}
+		// Find out if these are used by anything, but it looks like we should find a way to silence TestSummaries
+		// of type flakeSummary.
+		filePath := filepath.Join(framework.TestContext.ReportDir, fmt.Sprintf("%s_%s_%s.json", summary.SummaryKind(), filesystemSafeName(testName), now.Format(time.RFC3339)))
 		if err := ioutil.WriteFile(filePath, []byte(summary.PrintJSON()), 0644); err != nil {
 			fmt.Fprintf(os.Stderr, "error: Failed to write file %v with test data: %v\n", filePath, err)
 		}
 	}
 
 	if r == nil {
+		// Test can be considered successful, but may have flaked, see below:
+		ts.TestCases = append(ts.TestCases, &junitapi.JUnitTestCase{
+			Name:      testName,
+			Classname: className,
+			Duration:  testDuration,
+		})
 		if f != nil {
-			if message, ok := hasFrameworkFlake(f); ok {
-				tc.Failures = append(tc.Failures, &junit.Failure{
-					Type:    "Flake",
-					Message: message,
-					Value:   message,
+			if message, hasFlake := hasFrameworkFlake(f); hasFlake {
+				// Add another test case with the failure, this is a flake as we already added the success testcase above:
+				ts.TestCases = append(ts.TestCases, &junitapi.JUnitTestCase{
+					Name:          testName,
+					Classname:     className,
+					Duration:      testDuration,
+					FailureOutput: &junitapi.FailureOutput{Output: message},
 				})
 			}
 		}
 		return
 	}
 	framework.Logf("recover: %v", r)
 
+	testResult := &junitapi.JUnitTestCase{
+		Name:      testName,
+		Classname: className,
+		Duration:  testDuration,
+	}
 	switch r := r.(type) {
 	case ginkgowrapper.FailurePanic:
-		tc.Failures = []*junit.Failure{
-			{
-				Message: r.Message,
-				Type:    "Failure",
-				Value:   fmt.Sprintf("%s\n\n%s", r.Message, r.FullStackTrace),
-			},
-		}
+		testResult.FailureOutput = &junitapi.FailureOutput{Message: fmt.Sprintf("%s\n\n%s", r.Message, r.FullStackTrace)}
 	case e2eskipper.SkipPanic:
-		tc.Skipped = fmt.Sprintf("%s:%d %q", r.Filename, r.Line, r.Message)
+		testResult.SkipMessage = &junitapi.SkipMessage{Message: fmt.Sprintf("%s:%d %q", r.Filename, r.Line, r.Message)}
 	default:
-		tc.Errors = []*junit.Error{
-			{
-				Message: "Ginkgo panic encountered. See CDATA for details.",
-				Type:    "Panic",
-				Value:   fmt.Sprintf("%v\n\n%s", r, debug.Stack()),
-			},
-		}
+		testResult.FailureOutput = &junitapi.FailureOutput{Message: fmt.Sprintf("%v\n\n%s", r, debug.Stack())}
```

Comment:
This ran into @xueqzhan 's recent fix for the bad XML. I'm hoping this is ok still, my understanding we needed to keep the bad characters out of the message="" XML param. Our normal test cases do not use a message param, so this is gone entirely, and the full failure now resides within <failure/> tag.